### PR TITLE
Fixes #301 usbmux pair record plist parsing

### LINKED ISSUE #301: Fruity usbmux assumes ReadPairRecord response is an XML plist
I've had a persistent problem getting Frida's dynamic gadget injection working for iOS and the problem turns out to be that on my device, at least ( iPad6,11, iOS 13.4 but also at 13.3.1 ), the ReadPairRecord response is a binary plist.

The problem line is https://github.com/frida/frida-core/blob/master/src/fruity/usbmux.vala#L159, which assumes an XML plist.

For a quick test I made a hash of Plist.vala to add a "from_data(uint8[] data, uint8 fmt)" method which defaults to an "auto detect" by looking for the "bplist" magic for binary, used that in the problem area and it working.

I don't know vala so my code is a mess, but I can provide a PR if needed.

more data:
sniffing `/var/run/usbmux` you can see this for my device...
```
--
> 2020/04/08 15:59:56.857627  length=517 from=0 to=516
 05 02 00 00 01 00 00 00 08 00 00 00 01 00 00 00  ................
 3c 3f 78 6d 6c 20 76 65 72 73 69 6f 6e 3d 22 31  <?xml version="1
 2e 30 22 20 65 6e 63 6f 64 69 6e 67 3d 22 55 54  .0" encoding="UT
 46 2d 38 22 3f 3e 0a                             F-8"?>.
 3c 21 44 4f 43 54 59 50 45 20 70 6c 69 73 74 20  <!DOCTYPE plist 
 50 55 42 4c 49 43 20 22 2d 2f 2f 41 70 70 6c 65  PUBLIC "-//Apple
 2f 2f 44 54 44 20 50 4c 49 53 54 20 31 2e 30 2f  //DTD PLIST 1.0/
 2f 45 4e 22 20 22 68 74 74 70 3a 2f 2f 77 77 77  /EN" "http://www
 2e 61 70 70 6c 65 2e 63 6f 6d 2f 44 54 44 73 2f  .apple.com/DTDs/
 50 72 6f 70 65 72 74 79 4c 69 73 74 2d 31 2e 30  PropertyList-1.0
 2e 64 74 64 22 3e 0a                             .dtd">.
 3c 70 6c 69 73 74 20 76 65 72 73 69 6f 6e 3d 22  <plist version="
 31 2e 30 22 3e 0a                                1.0">.
 3c 64 69 63 74 3e 0a                             <dict>.
 09 3c 6b 65 79 3e 42 75 6e 64 6c 65 49 44 3c 2f  .<key>BundleID</
 6b 65 79 3e 0a                                   key>.
 09 3c 73 74 72 69 6e 67 3e 63 6f 6d 2e 61 70 70  .<string>com.app
 6c 65 2e 64 74 2e 58 63 6f 64 65 3c 2f 73 74 72  le.dt.Xcode</str
 69 6e 67 3e 0a                                   ing>.
 09 3c 6b 65 79 3e 43 6c 69 65 6e 74 56 65 72 73  .<key>ClientVers
 69 6f 6e 53 74 72 69 6e 67 3c 2f 6b 65 79 3e 0a  ionString</key>.
 09 3c 73 74 72 69 6e 67 3e 75 73 62 6d 75 78 64  .<string>usbmuxd
 2d 34 32 33 2e 35 30 2e 32 30 34 3c 2f 73 74 72  -423.50.204</str
 69 6e 67 3e 0a                                   ing>.
 09 3c 6b 65 79 3e 4d 65 73 73 61 67 65 54 79 70  .<key>MessageTyp
 65 3c 2f 6b 65 79 3e 0a                          e</key>.
 09 3c 73 74 72 69 6e 67 3e 52 65 61 64 50 61 69  .<string>ReadPai
 72 52 65 63 6f 72 64 3c 2f 73 74 72 69 6e 67 3e  rRecord</string>
 0a                                               .
 09 3c 6b 65 79 3e 50 61 69 72 52 65 63 6f 72 64  .<key>PairRecord
 49 44 3c 2f 6b 65 79 3e 0a                       ID</key>.
 09 3c 73 74 72 69 6e 67 3e 63 36 30 64 39 37 63  .<string>c60d97c
 62 61 64 32 66 31 38 34 66 64 62 31 62 65 61 31  bad2f184fdb1bea1
 61 31 39 37 33 38 65 36 64 64 38 66 39 37 31 39  a19738e6dd8f9719
 65 3c 2f 73 74 72 69 6e 67 3e 0a                 e</string>.
 09 3c 6b 65 79 3e 50 72 6f 67 4e 61 6d 65 3c 2f  .<key>ProgName</
 6b 65 79 3e 0a                                   key>.
 09 3c 73 74 72 69 6e 67 3e 58 63 6f 64 65 3c 2f  .<string>Xcode</
 73 74 72 69 6e 67 3e 0a                          string>.
 3c 2f 64 69 63 74 3e 0a                          </dict>.
 3c 2f 70 6c 69 73 74 3e 0a                       </plist>.

-

// PairRecord data is a binary plist...
< 2020/04/08 15:59:56.861613  length=8192 from=0 to=8191
 f0 24 00 00 01 00 00 00 08 00 00 00 01 00 00 00  .$..............
 3c 3f 78 6d 6c 20 76 65 72 73 69 6f 6e 3d 22 31  <?xml version="1
 2e 30 22 20 65 6e 63 6f 64 69 6e 67 3d 22 55 54  .0" encoding="UT
 46 2d 38 22 3f 3e 0a                             F-8"?>.
 3c 21 44 4f 43 54 59 50 45 20 70 6c 69 73 74 20  <!DOCTYPE plist 
 50 55 42 4c 49 43 20 22 2d 2f 2f 41 70 70 6c 65  PUBLIC "-//Apple
 2f 2f 44 54 44 20 50 4c 49 53 54 20 31 2e 30 2f  //DTD PLIST 1.0/
 2f 45 4e 22 20 22 68 74 74 70 3a 2f 2f 77 77 77  /EN" "http://www
 2e 61 70 70 6c 65 2e 63 6f 6d 2f 44 54 44 73 2f  .apple.com/DTDs/
 50 72 6f 70 65 72 74 79 4c 69 73 74 2d 31 2e 30  PropertyList-1.0
 2e 64 74 64 22 3e 0a                             .dtd">.
 3c 70 6c 69 73 74 20 76 65 72 73 69 6f 6e 3d 22  <plist version="
 31 2e 30 22 3e 0a                                1.0">.
 3c 64 69 63 74 3e 0a                             <dict>.
 09 3c 6b 65 79 3e 50 61 69 72 52 65 63 6f 72 64  .<key>PairRecord
 44 61 74 61 3c 2f 6b 65 79 3e 0a                 Data</key>.
 09 3c 64 61 74 61 3e 0a                          .<data>.
 09 59 6e 42 73 61 58 4e 30 4d 44 44 5a 41 51 4d  .YnBsaXN0MDDZAQM
 46 42 77 6b 4c 44 51 38 52 41 67 51 47 43 41 6f  FBwkLDQ8RAgQGCAo
 4d 44 68 41 53 58 78 41 52 52 47 56 32 61 57 4e  MDhASXxARRGV2aWN
 6c 51 32 56 79 64 47 6c 6d 61 57 4e 68 64 47 56  lQ2VydGlmaWNhdGV
 50 45 51 50 74 0a                                PEQPt.
 09 4c 53 30 74 4c 53 31 43 52 55 64 4a 54 69 42  .LS0tLS1CRUdJTiB
 44 52 56 4a 55 53 55 5a 4a 51 30 46 55 52 53 30  DRVJUSUZJQ0FURS0
 74 4c 53 30 74 43 6b 31 4a 53 55 4e 31 61 6b 4e  tLS0tCk1JSUN1akN
 44 51 57 46 4c 5a 30 46 33 53 55 4a 42 5a 30 6c  DQWFLZ0F3SUJBZ0l
 43 51 55 52 42 0a                                CQURB.
// ...
```

### PR DESCRIPTION
**Pull request by pachoo**
* added Plist.from_data to support auto-detecting plist format.
* UsbmuxClient.read_pair_record() supports auto-detecting plist format

### CODE REVIEW

**Comment by pachoo** [comment]
first time with vala, hopefully the structure and memory management is appropriate :)

File: src/fruity/plist.vala [hunk 1]
```diff
@@ -1,15 +1,37 @@
 namespace Frida.Fruity {
 	public class Plist : PlistDict {
+		public const uint8 FMT_AUTO = 0;
+		public const uint8 FMT_BIN = 1;
+		public const uint8 FMT_XML = 2;
```
Comment (oleavr) [inline src/fruity/plist.vala:5]: Would be great to use an enum here. Something like:

```vala
		public enum Format {
			AUTO,
			BINARY,
			XML
		}
```

File: src/fruity/plist.vala [hunk 2]
```diff
@@ -1,15 +1,37 @@
...
+			this.from_data (xml.data, FMT_XML);
+		}
+
+		public Plist.from_data (uint8[] data, uint8 fmt=FMT_AUTO) throws PlistError {
+			if (fmt == FMT_AUTO) {
+				unowned string magic = (string) data;
+				if (magic.has_prefix ("bplist")) {
+					fmt = FMT_BIN;
+				} else {
+					// assume XML
```
Comment (oleavr) [inline src/fruity/plist.vala:22]:
```suggestion
```

Let's drop this comment as it doesn't add anything. More details on the style [here](https://github.com/frida/frida/blob/master/CONTRIBUTING.md).

File: src/fruity/plist.vala [hunk 3]
```diff
@@ -1,15 +1,37 @@
...
+				}
+			}
+			if (fmt == FMT_BIN) {
+				var parser = new BinaryParser (this);
+				parser.parse (data);
+			} else if (fmt == FMT_XML) {
+				var parser = new XmlParser (this);
+				parser.parse ((string)data);
+			} else {
+				throw new PlistError.INVALID_DATA ("Invalid format specified: %d", fmt);
```
Comment (oleavr) [inline src/fruity/plist.vala:33]: Using an enum, this should be `assert_not_reached ()` instead.

File: src/fruity/usbmux.vala [hunk 4]
```diff
@@ -154,9 +154,8 @@ namespace Frida.Fruity {
 				}
 
 				var raw_record = response.get_bytes ("PairRecordData");
-				unowned string record_xml_unterminated = (string) raw_record.get_data ();
-				string record_xml = record_xml_unterminated[0:raw_record.length];
-				return new Plist.from_xml (record_xml);
+				var record_data = raw_record.get_data ();
+				return new Plist.from_data (record_data);
```
Comment (oleavr) [inline src/fruity/usbmux.vala:158]:
```suggestion
				return new Plist.from_data (raw_record.get_data ());
```

File: src/fruity/plist.vala [hunk 5]
```diff
@@ -1,15 +1,37 @@
...
+					// assume XML
+					fmt = FMT_XML;
+				}
+			}
+			if (fmt == FMT_BIN) {
+				var parser = new BinaryParser (this);
+				parser.parse (data);
+			} else if (fmt == FMT_XML) {
+				var parser = new XmlParser (this);
+				parser.parse ((string)data);
```
Comment (oleavr) [inline src/fruity/plist.vala:31]:
```suggestion
				parser.parse ((string) data);
```

File: src/fruity/plist.vala [hunk 6]
```diff
@@ -1,15 +1,37 @@
...
+			this.from_data (data, FMT_BIN);
 		}
 
 		public Plist.from_xml (string xml) throws PlistError {
-			var parser = new XmlParser (this);
-			parser.parse (xml);
+			this.from_data (xml.data, FMT_XML);
+		}
+
+		public Plist.from_data (uint8[] data, uint8 fmt=FMT_AUTO) throws PlistError {
```
Comment (oleavr) [inline src/fruity/plist.vala:16]:
```suggestion
		public Plist.from_data (uint8[] data, uint8 fmt = FMT_AUTO) throws PlistError {
```

(Forgot to say that for default arguments, `=` should have one space on each side.)

File: src/fruity/plist.vala [hunk 7]
```diff
@@ -1,36 +1,37 @@
 namespace Frida.Fruity {
 	public class Plist : PlistDict {
-		public const uint8 FMT_AUTO = 0;
-		public const uint8 FMT_BIN = 1;
-		public const uint8 FMT_XML = 2;
+		public enum Format {
+			AUTO,
+			BINARY,
+			XML
+		}
```
Comment (oleavr) [inline src/fruity/plist.vala:7]:
```suggestion
		}

```

File: src/fruity/plist.vala [hunk 8]
```diff
@@ -1,36 +1,37 @@
...
 		}
 
 		public Plist.from_xml (string xml) throws PlistError {
-			this.from_data (xml.data, FMT_XML);
+			this.from_data (xml.data, XML);
 		}
 
-		public Plist.from_data (uint8[] data, uint8 fmt=FMT_AUTO) throws PlistError {
-			if (fmt == FMT_AUTO) {
+		public Plist.from_data (uint8[] data, Format fmt = AUTO) throws PlistError {
```
Comment (oleavr) [inline src/fruity/plist.vala:18]:
```suggestion
		public Plist.from_data (uint8[] data, Format format = AUTO) throws PlistError {

```

File: src/fruity/plist.vala (hunk 7 above)
Comment (pachoo) [reply src/fruity/plist.vala:7]: newline added to line 8 in 1d1de403be669436eeb3eaa0cb79ac62a9e89042

File: src/fruity/plist.vala (hunk 8 above)
Comment (pachoo) [reply src/fruity/plist.vala]: parameter fmt renamed to format in 1d1de403be669436eeb3eaa0cb79ac62a9e89042

**Pull request merged**